### ISSUE TITLE
Inner exceptions on @property DocType

### ISSUE THREAD

**Issue by dourvaris** [elasticsearch/elasticsearch-dsl-py#55]
Using a property on classes makes debugging a nightmare, for example:
  
```python
from elasticsearch_dsl import DocType, Search
class Foo(DocType):
    @property
    def bars(self):
        return Search().eecute() # with a filter in reality

foo = Foo()
foo.bars
```
This will raise **AttributeError: 'Foo' object has no attribute 'bars'**

The actual exception is **KeyError: "There is no connection with alias 'default'."**
```
  File /elasticsearch-dsl/elasticsearch_dsl/search.py", line 366, in execute
    es = connections.get_connection(self._using)
  File "/elasticsearch-dsl/elasticsearch_dsl/connections.py", line 53, in get_connection
    raise KeyError('There is no connection with alias %r.' % alias)
```

I think the except AttributeError clauses in AttrDict and ObjectBase should be removed, I don't see how they add anything.

**Comment by dourvaris** [comment]
Another problem with properties:
```python
class Foo(DocType):
     _value = 'old'
     @property
     def value(self):
         if self._value is 'old':
             self._value = 'new'
         return self._value
 
Foo().value
>>> 'old'
```

**Comment by HonzaKral** [comment]
They add cosistency - if I am accesing an attribute: `result.title` and the property doesn't exist, I expect an `AttributeError`, not a `KeyError`.

I can see the issues in the properties you wrote here, but I don't see the connection with `AttrDict` and `ObjectBase`, could you please elaborate where these are hurting you? Thanks

**Comment by dourvaris** [comment]
I agree on the consistency aspect, and would definitely expect an `AttributeError` on attribute access, I don't have a good answer there yet, maybe an `AttrDictAttributeError` which will keep the original exception.. I'm not sure if it's possible.

The problem is due to the nature of attrdicts having the tendancy to consider attribute failure as a _missing attribute_ rather than an actual exception. This makes the first access of an attribute on an attrdict the initial error. 

The places where this will bite you in the ass on a complicated model are plenty, a simple case:

```python
from elasticsearch_dsl import DocType
class Product(DocType):
    pass

class Category(DocType):
    @property
    def products(self):
        return Product.all()

category = Category(name='laptop')
category.products
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dan/env/local/lib/python2.7/site-packages/elasticsearch_dsl/document.py", line 166, in __getattr__
    return super(DocType, self).__getattr__(name)
  File "/home/dan/env/local/lib/python2.7/site-packages/elasticsearch_dsl/utils.py", line 374, in __getattr__
    return super(ObjectBase, self).__getattr__(name)
  File "/home/dan/env/local/lib/python2.7/site-packages/elasticsearch_dsl/utils.py", line 81, in __getattr__
    '%r object has no attribute %r' % (self.__class__.__name__, attr_name))
AttributeError: 'Category' object has no attribute 'products'
```

Category *does* have an attribute products, `Product.all` is missing - but the exception will make no mention of this. Consider a complex model with methods/attributes all depending on each other. The result is wasting time following the chain of events, which is what we have tracebacks for in the first place.

**Comment by HonzaKral** [comment]
Oh, I see - the problem is that when your property raises an `AttributeError` then `AttrDict.__getattr__` is called and it will raise it's own `AttributeError` not mentioning the original (because it cannot).

This is actually a known Python behavior (you might even call it a bug) that currently has no workaround. The consensus amongst python people is to just make sure your properties don't raise `AttributeError` - that's what's causing `__getattr__` to be invoked.

Why this cannot be fixed is because there are legitimate use cases for this behavior - when raising `AttributeError` from a property should trigger a `__getattr__` call - it's an option left there for the programmer. If you don't wish it it's recommended you catch `AttributeError` and raise a different error instead, that will be correctly propagated.

As far as I know none of the properties defined by this library can ever raise `AttributeError` unless something goes seriously wrong (people messing with our objects' internal states etc). If you find a place where it is not so we will definitely treat it as a bug and fix it asap.

**Comment by dourvaris** [comment]
I'm not talking about properties intentionally raising an AttributeError. I'm saying that there will be times when a simple typo bug will lead to an AttributeError being raised, which will be hidden because ObjectBase try: excepts: it away.

```
>>> from elasticsearch_dsl import DocType
>>> 
>>> class RegularClass(object):
...     @property
...     def foo(self):
...         return self.typo
... 
>>> RegularClass().foo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in foo
AttributeError: 'RegularClass' object has no attribute 'typo'
>>> 
>>> class DocTypeClass(DocType):
...     @property
...     def foo(self):
...         return self.typo
... 
>>> DocTypeClass().foo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dan/env/local/lib/python2.7/site-packages/elasticsearch_dsl/document.py", line 166, in __getattr__
    return super(DocType, self).__getattr__(name)
  File "/home/dan/env/local/lib/python2.7/site-packages/elasticsearch_dsl/utils.py", line 374, in __getattr__
    return super(ObjectBase, self).__getattr__(name)
  File "/home/dan/env/local/lib/python2.7/site-packages/elasticsearch_dsl/utils.py", line 81, in __getattr__
    '%r object has no attribute %r' % (self.__class__.__name__, attr_name))
AttributeError: 'DocTypeClass' object has no attribute 'foo'
```

You'd expect an object in python to tell you that offending line is `return self.typo` not that the object has no attribute foo.

**Comment by HonzaKral** [comment]
Just to be clear, `ObjectBase` does not try/except it away - it's python's `object` class that does this in it's `__getattribute__` - when a class has `__getattr__` defined, this is the behavior:

    class A(object):
        @property
        def f(self):
            return self.g

        def __getattr__(self, name):
            raise AttributeError(name + '!!!')

    print(A().f)

it is standard python behavior and that's why I am reluctant to "fix" it. The fix would also contain modifying the object's `__getattribute__` instead of `__getattr__` which is potentially dangerous and can have some performance implications. I will run some tests and see what the difference would be if it's worth it.

**Comment by HonzaKral** [comment]
Switching over to `__getattribute__` will make attribute access 3x slower as it currently is with `__getattr__` if the attribute exists in the underlying dict and 2x slower if it doesn't. We are already much slower that raw attribute access so I'd hate to compound the issue.

I understand the frustration with misleading errors, I am just not sure it's worth the slowdown, especially when it's just an unfortunate feature of python and not behavior we are introducing.

**Comment by HonzaKral** [comment]
I am sorry, my assumptions that it won't happen often has been shown as invalid. Thank you for raising this, I will fix this as part of #81
